### PR TITLE
Rework predictions, a hint of flight plan

### PR DESCRIPTION
**Pull request by eggrobin**
Rework the logic of the predictions (they're integrated by `Vessel` now), a potential implementation of the flight plan, with no interface.

### CODE REVIEW

File: ksp_plugin/plugin.cpp [hunk 1]
```diff
@@ -295,6 +294,29 @@ RelativeDegreesOfFreedom<AliceSun> Plugin::CelestialFromParent(
   return result;
 }
 
+void Plugin::UpdatePrediction(GUID const & vessel_guid) const {
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: s/ &/&/

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -318,31 +340,50 @@ RenderedTrajectory<World> Plugin::RenderedVesselTrajectory(
                           sun_world_position);
 }
 
+int Plugin::FlightPlanSize(GUID const & vessel_guid) const {
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: s/ &/&/

File: ksp_plugin/plugin.cpp [hunk 3]
```diff
@@ -318,31 +340,50 @@ RenderedTrajectory<World> Plugin::RenderedVesselTrajectory(
...
+  CHECK(!initializing_);
+  return find_vessel_by_guid_or_die(vessel_guid)->flight_plan().size();
+}
+
+bool Plugin::HasPrediction(GUID const & vessel_guid) const {
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Same.

File: ksp_plugin/plugin.cpp [hunk 4]
```diff
@@ -318,31 +340,50 @@ RenderedTrajectory<World> Plugin::RenderedVesselTrajectory(
...
-  DeletePredictions();
-  predicted_vessel_ = nullptr;
+RenderedTrajectory<World> Plugin::RenderedFlightPlan(
+    GUID const& vessel_guid,
+    int const plan_phase,
+    not_null<RenderingTransforms*> const transforms,
+    Position<World> const& sun_world_position) {
+  CHECK(!initializing_);
+  Vessel const& vessel = *find_vessel_by_guid_or_die(vessel_guid);
+  CHECK_GT(vessel.flight_plan().size(), plan_phase);
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Preference for inverting: CHECH_LT(plan_phase, ...).

File: ksp_plugin/vessel.hpp [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
 #include "physics/discrete_trajectory.hpp"
+#include "physics/ephemeris.hpp"
 #include "physics/massless_body.hpp"
 #include "quantities/named_quantities.hpp"
 #include "serialization/ksp_plugin.pb.h"
 
 namespace principia {
 
+using physics::Ephemeris;
 using physics::DiscreteTrajectory;
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: Alphabetize.

File: ksp_plugin/vessel.hpp [hunk 6]
```diff
@@ -83,11 +96,37 @@ class Vessel {
...
-  // Deletes the |prediction_|.
+  // Fills |flight_plan_| with predictions using the given |ephemeris| for
+  // successive manœuvres, with the given prediction tolerances for the coasting
+  // phases, and the given prolongation tolerances for the manœuvres.  Uses the
+  // given |integrator|.
+  // Deletes any pre-existing predictions.
+  // Does nothing unless |is_synchronized()|, pending the removal of
+  // synchronization.
+  // TODO(egg): struct containing (integrator, length tol, speed tol) so we
+  // don't need that many parameters...
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: +1

File: ksp_plugin/vessel_body.hpp [hunk 7]
```diff
@@ -58,20 +60,33 @@ Vessel::mutable_prolongation() {
   return prolongation_;
 }
 
-inline DiscreteTrajectory<Barycentric> const& Vessel::prediction() const {
+inline std::vector<not_null<DiscreteTrajectory<Barycentric>*>> const&
```
Comment (pleroy) [inline ksp_plugin/vessel_body.hpp]: Remove these inline.

File: ksp_plugin/vessel_body.hpp [hunk 8]
```diff
@@ -101,14 +116,88 @@ inline void Vessel::ResetProlongation(Instant const& time) {
...
+  }
+  DeleteFlightPlan();
+  flight_plan_.emplace_back(
+      mutable_history()->NewForkWithCopy(history().last().time()));
+  // If prolongation has no additional points this will do nothing (although it
+  // might warn).
+  flight_plan_.back()->Append(prolongation().last().time(),
+                              prolongation().last().degrees_of_freedom());
+  for (auto const& manœuvre : manœuvres_) {
+    not_null<DiscreteTrajectory<Barycentric>*> coast_trajectory =
```
Comment (pleroy) [inline ksp_plugin/vessel_body.hpp]: const?

File: ksp_plugin/vessel_body.hpp [hunk 9]
```diff
@@ -101,14 +116,88 @@ inline void Vessel::ResetProlongation(Instant const& time) {
...
+        flight_plan_.back();
+    ephemeris->FlowWithAdaptiveStep(coast_trajectory,
+                                    prediction_length_tolerance,
+                                    prediction_speed_tolerance,
+                                    integrator,
+                                    manœuvre->initial_time());
+    flight_plan_.emplace_back(
+        coast_trajectory->NewForkWithCopy(coast_trajectory->last().time()));
+    not_null<DiscreteTrajectory<Barycentric>*> burn_trajectory =
```
Comment (pleroy) [inline ksp_plugin/vessel_body.hpp]: const?

File: ksp_plugin_adapter/ksp_plugin_interface.cs [hunk 10]
```diff
@@ -248,10 +248,25 @@ struct KSPPart {
       XYZ sun_world_position);
 
   [DllImport(dllName           : kDllPath,
+             EntryPoint        = "principia__UpdatePrediction",
+             CallingConvention = CallingConvention.Cdecl)]
+  internal static extern void UpdatePrediction(
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_interface.cs]: private?

File: ksp_plugin/plugin.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: doen.

File: ksp_plugin/plugin.cpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: done.

File: ksp_plugin/plugin.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: done.

File: ksp_plugin/plugin.cpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Agreed.

File: ksp_plugin/vessel.hpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/vessel.hpp]: Done.

File: ksp_plugin/vessel_body.hpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin/vessel_body.hpp]: Pinsule.

File: ksp_plugin/vessel_body.hpp (hunk 8 above)
Comment (eggrobin) [inline ksp_plugin/vessel_body.hpp]: Done.

File: ksp_plugin/vessel_body.hpp (hunk 9 above)
Comment (eggrobin) [inline ksp_plugin/vessel_body.hpp]: Dnoe.

File: ksp_plugin_adapter/ksp_plugin_interface.cs (hunk 10 above)
Comment (eggrobin) [inline ksp_plugin_adapter/ksp_plugin_interface.cs]: Done.

File: ksp_plugin/vessel.hpp (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin/vessel.hpp]: #657.

**Pull request merged**